File: src/backend/geo_lib/validation/coordinate_validation.py
"""
Coordinate validation utilities for GeoJSON coordinates.

This module provides validation functions for coordinate arrays, ensuring they:
1. Match the expected structure for the geometry type
2. Are within valid bounds (lon: -180 to 180, lat: -90 to 90)
3. Are not swapped (lat/lon order)
"""

from typing import List, Any, Tuple, Optional


class CoordinateValidationError(Exception):
    """Exception raised when coordinate validation fails."""
    pass


def _validate_point_coordinate(point: List[Any]) -> Tuple[float, float]:
    """
    Validate a single point coordinate [lon, lat] or [lon, lat, elevation].
    
    Args:
        point: Coordinate array [lon, lat] or [lon, lat, elevation]
        
    Returns:
        Tuple of (lon, lat)
        
    Raises:
        CoordinateValidationError: If coordinate is invalid
    """
    if not isinstance(point, list):
        raise CoordinateValidationError(f"Point coordinate must be an array, got {type(point)}")
    
    if len(point) < 2:
        raise CoordinateValidationError(f"Point coordinate must have at least 2 elements [lon, lat], got {len(point)}")
    
    lon = point[0]
    lat = point[1]
    
    # Check types
    if not isinstance(lon, (int, float)):
        raise CoordinateValidationError(f"Longitude must be a number, got {type(lon)}")
    if not isinstance(lat, (int, float)):
        raise CoordinateValidationError(f"Latitude must be a number, got {type(lat)}")
    
    # Check for None values
    if lon is None or lat is None:
        raise CoordinateValidationError("Coordinate contains None values")
    
    # Convert to float for bounds checking
    try:
        lon = float(lon)
        lat = float(lat)
    except (ValueError, TypeError, OverflowError):
        raise CoordinateValidationError("Coordinate values must be valid numbers")
    
    # Check for NaN or Infinity
    import math
    if math.isnan(lon) or math.isnan(lat):
        raise CoordinateValidationError("Coordinate values cannot be NaN")
    if math.isinf(lon) or math.isinf(lat):
        raise CoordinateValidationError("Coordinate values cannot be Infinity")
    
    # Check bounds
    if not (-180 <= lon <= 180):
        raise CoordinateValidationError(f"Longitude {lon} is out of bounds [-180, 180]")
    if not (-90 <= lat <= 90):
        raise CoordinateValidationError(f"Latitude {lat} is out of bounds [-90, 90]")
    
    # Detect lat/lon swapping
    # Only flag as swapped if coordinates are clearly wrong
    # Don't flag valid longitudes just because abs(lon) > 90 (many valid longitudes are > 90)
    
    # Case 1: Latitude is clearly out of bounds (abs(lat) > 90)
    # This means the value in the lat position is actually a longitude
    if abs(lat) > 90:
        raise CoordinateValidationError(
            f"Coordinates appear to be swapped. Latitude {lat} is outside valid range [-90, 90]."
        )
    
    # Note: We don't check for lon > 180 here because that's already caught by bounds check above
    # and would fail before we get to swap detection
    
    # Case 2: Both values are in valid ranges, but pattern suggests a swap
    # Only flag if lon is in lat range (-90 to 90) AND lat is clearly a longitude value
    # This catches cases like [40, -120] where 40 could be lat but -120 is clearly a lon
    # But we've already validated that abs(lat) <= 90, so if we get here, both are valid
    # We can't reliably detect swaps when both values are in valid ranges
    # So we only flag obvious cases that passed bounds but are clearly wrong
    # (This case is now empty since bounds check handles everything)
    
    return (lon, lat)


def _check_multiple_points_for_swap(points: List[Tuple[float, float]]) -> Optional[str]:
    """
    Check multiple points to detect consistent lat/lon swapping pattern.
    
    Args:
        points: List of (lon, lat) tuples
        
    Returns:
        Error message if swap detected, None otherwise
    """
    if len(points) < 2:
        return None
    
    # Count how many points look swapped
    swap_count = 0
    total_checked = 0
    
    for lon, lat in points:
        # Skip if either is clearly out of bounds (already caught by _validate_point_coordinate)
        if abs(lon) > 180 or abs(lat) > 90:
            continue
        
        total_checked += 1
        
        # If both are in valid ranges but lon is in lat range and lat is in lon range
        if (-90 <= lon <= 90) and (-180 <= lat <= 180):
            # Check if this looks like a swap
            if abs(lon) > abs(lat):
                swap_count += 1
    
    # If majority of points look swapped, report it
    if total_checked >= 2 and swap_count > total_checked * 0.5:
        return (
            f"Multiple coordinates appear to be swapped. "
            f"Expected [longitude, latitude] format. "
            f"Longitude should be first (range -180 to 180), latitude second (range -90 to 90)."
        )
    
    return None


def _validate_coordinates_structure(coordinates: Any, expected_depth: int, geometry_type: str) -> List[Tuple[float, float]]:
    """
    Validate coordinate structure and depth, returning all points for swap detection.
    
    Args:
        coordinates: Coordinate array
        expected_depth: Expected nesting depth (0 for Point, 1 for LineString/MultiPoint, etc.)
        geometry_type: Geometry type name for error messages
        
    Returns:
        List of (lon, lat) tuples from all points
        
    Raises:
        CoordinateValidationError: If structure is invalid
    """
    if not isinstance(coordinates, list):
        raise CoordinateValidationError(f"{geometry_type} coordinates must be an array")
    
    if expected_depth == 0:
        # Point: [lon, lat] or [lon, lat, elevation]
        lon, lat = _validate_point_coordinate(coordinates)
        return [(lon, lat)]
    
    elif expected_depth == 1:
        # LineString or MultiPoint: [[lon, lat], ...]
        if len(coordinates) == 0:
            raise CoordinateValidationError(f"{geometry_type} must have at least one coordinate")
        
        points = []
        for i, point in enumerate(coordinates):
            try:
                lon, lat = _validate_point_coordinate(point)
                points.append((lon, lat))
            except CoordinateValidationError as e:
                raise CoordinateValidationError(f"Invalid coordinate at index {i}: {str(e)}")
        
        return points
    
    elif expected_depth == 2:
        # Polygon or MultiLineString: [[[lon, lat], ...], ...]
        if len(coordinates) == 0:
            raise CoordinateValidationError(f"{geometry_type} must have at least one ring/line")
        
        all_points = []
        for ring_idx, ring in enumerate(coordinates):
            if not isinstance(ring, list):
                raise CoordinateValidationError(f"{geometry_type} ring/line at index {ring_idx} must be an array")
            if len(ring) == 0:
                raise CoordinateValidationError(f"{geometry_type} ring/line at index {ring_idx} must have at least one coordinate")
            
            for point_idx, point in enumerate(ring):
                try:
                    lon, lat = _validate_point_coordinate(point)
                    all_points.append((lon, lat))
                except CoordinateValidationError as e:
                    raise CoordinateValidationError(
                        f"Invalid coordinate at ring/line {ring_idx}, point {point_idx}: {str(e)}"
                    )
        
        return all_points
    
    elif expected_depth == 3:
        # MultiPolygon: [[[[lon, lat], ...], ...], ...]
        if len(coordinates) == 0:
            raise CoordinateValidationError(f"{geometry_type} must have at least one polygon")
        
        all_points = []
        for poly_idx, polygon in enumerate(coordinates):
            if not isinstance(polygon, list):
                raise CoordinateValidationError(f"{geometry_type} polygon at index {poly_idx} must be an array")
            if len(polygon) == 0:
                raise CoordinateValidationError(f"{geometry_type} polygon at index {poly_idx} must have at least one ring")
            
            for ring_idx, ring in enumerate(polygon):
                if not isinstance(ring, list):
                    raise CoordinateValidationError(
                        f"{geometry_type} polygon {poly_idx}, ring {ring_idx} must be an array"
                    )
                if len(ring) == 0:
                    raise CoordinateValidationError(
                        f"{geometry_type} polygon {poly_idx}, ring {ring_idx} must have at least one coordinate"
                    )
                
                for point_idx, point in enumerate(ring):
                    try:
                        lon, lat = _validate_point_coordinate(point)
                        all_points.append((lon, lat))
                    except CoordinateValidationError as e:
                        raise CoordinateValidationError(
                            f"Invalid coordinate at polygon {poly_idx}, ring {ring_idx}, point {point_idx}: {str(e)}"
                        )
        
        return all_points
    
    else:
        raise CoordinateValidationError(f"Unsupported coordinate depth: {expected_depth}")


def validate_coordinates_for_geometry_type(coordinates: Any, geometry_type: str) -> None:
    """
    Validate coordinates array matches the expected structure for the geometry type.
    
    Validates:
    1. Coordinates are not None or empty
    2. Structure depth matches geometry type
    3. All coordinates are within bounds (lon: -180 to 180, lat: -90 to 90)
    4. Coordinates are not swapped (lat/lon order)
    
    Args:
        coordinates: Coordinate array to validate
        geometry_type: Geometry type (Point, LineString, Polygon, MultiPoint, MultiLineString, MultiPolygon)
        
    Raises:
        CoordinateValidationError: If validation fails with descriptive error message
    """
    # Reject None or empty coordinates
    if coordinates is None:
        raise CoordinateValidationError("Coordinates cannot be null or empty")
    
    if not isinstance(coordinates, list):
        raise CoordinateValidationError(f"Coordinates must be an array, got {type(coordinates)}")
    
    geometry_type_lower = geometry_type.lower()
    
    # Determine expected depth
    depth_map = {
        'point': 0,
        'linestring': 1,
        'polygon': 2,
        'multipoint': 1,
        'multilinestring': 2,
        'multipolygon': 3,
    }
    
    if geometry_type_lower not in depth_map:
        raise CoordinateValidationError(f"Unsupported geometry type: {geometry_type}")
    
    expected_depth = depth_map[geometry_type_lower]
    
    # For Point, empty array is invalid (must have [lon, lat])
    if expected_depth == 0 and len(coordinates) == 0:
        raise CoordinateValidationError("Point coordinates cannot be empty. Must be [longitude, latitude]")
    
    # For other types, empty arrays are checked in _validate_coordinates_structure
    # but we check here too for early failure
    if expected_depth > 0 and len(coordinates) == 0:
        raise CoordinateValidationError(f"{geometry_type.capitalize()} coordinates cannot be empty")
    
    # Validate structure and get all points
    points = _validate_coordinates_structure(coordinates, expected_depth, geometry_type.capitalize())
    
    # Additional swap detection for multi-point geometries
    if len(points) > 1:
        swap_error = _check_multiple_points_for_swap(points)
        if swap_error:
            raise CoordinateValidationError(swap_error)

File: src/frontend/src/utils/coordinateValidation.js
/**
 * Coordinate validation utilities for GeoJSON coordinates.
 * 
 * Validates coordinate arrays to ensure they:
 * 1. Match the expected structure for the geometry type
 * 2. Are within valid bounds (lon: -180 to 180, lat: -90 to 90)
 * 3. Are not swapped (lat/lon order)
 */

/**
 * Validate a single point coordinate [lon, lat] or [lon, lat, elevation].
 * 
 * @param {Array} point - Coordinate array [lon, lat] or [lon, lat, elevation]
 * @returns {{lon: number, lat: number}} - Validated lon and lat
 * @throws {Error} - If coordinate is invalid
 */
function validatePointCoordinate(point) {
  if (!Array.isArray(point)) {
    throw new Error(`Point coordinate must be an array, got ${typeof point}`)
  }
  
  if (point.length < 2) {
    throw new Error(`Point coordinate must have at least 2 elements [lon, lat], got ${point.length}`)
  }
  
  const lon = point[0]
  const lat = point[1]
  
  // Check types
  if (typeof lon !== 'number' && typeof lon !== 'string') {
    throw new Error(`Longitude must be a number, got ${typeof lon}`)
  }
  if (typeof lat !== 'number' && typeof lat !== 'string') {
    throw new Error(`Latitude must be a number, got ${typeof lat}`)
  }
  
  // Check for null/undefined
  if (lon === null || lon === undefined || lat === null || lat === undefined) {
    throw new Error('Coordinate contains null or undefined values')
  }
  
  // Convert to number
  const lonNum = parseFloat(lon)
  const latNum = parseFloat(lat)
  
  if (isNaN(lonNum) || isNaN(latNum)) {
    throw new Error('Coordinate values must be valid numbers')
  }
  
  // Check for Infinity
  if (!isFinite(lonNum) || !isFinite(latNum)) {
    throw new Error('Coordinate values cannot be Infinity')
  }
  
  // Check bounds — flag obvious lat/lon swap before generic latitude out-of-bounds text
  if (!(-180 <= lonNum && lonNum <= 180)) {
    throw new Error(`Longitude ${lonNum} is out of bounds [-180, 180]`)
  }
  if (Math.abs(latNum) > 90) {
    throw new Error(
      `Coordinates appear to be swapped. Latitude ${latNum} is outside valid range [-90, 90].`
    )
  }
  if (!(-90 <= latNum && latNum <= 90)) {
    throw new Error(`Latitude ${latNum} is out of bounds [-90, 90]`)
  }

  return { lon: lonNum, lat: latNum }
}

/**
 * Check multiple points to detect consistent lat/lon swapping pattern.
 * 
 * @param {Array<{lon: number, lat: number}>} points - List of (lon, lat) objects
 * @returns {string|null} - Error message if swap detected, null otherwise
 */
function checkMultiplePointsForSwap(points) {
  if (points.length < 2) {
    return null
  }
  
  // Count how many points look swapped
  let swapCount = 0
  let totalChecked = 0
  
  for (const { lon, lat } of points) {
    // Skip if either is clearly out of bounds (already caught by validatePointCoordinate)
    if (Math.abs(lon) > 180 || Math.abs(lat) > 90) {
      continue
    }
    
    totalChecked += 1
    
    // If both are in valid ranges but lon is in lat range and lat is in lon range
    if ((-90 <= lon && lon <= 90) && (-180 <= lat && lat <= 180)) {
      // Check if this looks like a swap
      if (Math.abs(lon) > Math.abs(lat)) {
        swapCount += 1
      }
    }
  }
  
  // If majority of points look swapped, report it
  if (totalChecked >= 2 && swapCount > totalChecked * 0.5) {
    return (
      `Multiple coordinates appear to be swapped. ` +
      `Expected [longitude, latitude] format. ` +
      `Longitude should be first (range -180 to 180), latitude second (range -90 to 90).`
    )
  }
  
  return null
}

/**
 * Validate coordinate structure and depth, returning all points for swap detection.
 * 
 * @param {*} coordinates - Coordinate array
 * @param {number} expectedDepth - Expected nesting depth (0 for Point, 1 for LineString/MultiPoint, etc.)
 * @param {string} geometryType - Geometry type name for error messages
 * @returns {Array<{lon: number, lat: number}>} - List of (lon, lat) objects from all points
 * @throws {Error} - If structure is invalid
 */
function validateCoordinatesStructure(coordinates, expectedDepth, geometryType) {
  if (!Array.isArray(coordinates)) {
    throw new Error(`${geometryType} coordinates must be an array`)
  }
  
  if (expectedDepth === 0) {
    // Point: [lon, lat] or [lon, lat, elevation]
    const { lon, lat } = validatePointCoordinate(coordinates)
    return [{ lon, lat }]
  } else if (expectedDepth === 1) {
    // LineString or MultiPoint: [[lon, lat], ...]
    if (coordinates.length === 0) {
      throw new Error(`${geometryType} must have at least one coordinate`)
    }
    
    const points = []
    for (let i = 0; i < coordinates.length; i++) {
      try {
        const { lon, lat } = validatePointCoordinate(coordinates[i])
        points.push({ lon, lat })
      } catch (e) {
        throw new Error(`Invalid coordinate at index ${i}: ${e.message}`)
      }
    }
    
    return points
  } else if (expectedDepth === 2) {
    // Polygon or MultiLineString: [[[lon, lat], ...], ...]
    if (coordinates.length === 0) {
      throw new Error(`${geometryType} must have at least one ring/line`)
    }
    
    const allPoints = []
    for (let ringIdx = 0; ringIdx < coordinates.length; ringIdx++) {
      const ring = coordinates[ringIdx]
      if (!Array.isArray(ring)) {
        throw new Error(`${geometryType} ring/line at index ${ringIdx} must be an array`)
      }
      if (ring.length === 0) {
        throw new Error(`${geometryType} ring/line at index ${ringIdx} must have at least one coordinate`)
      }
      
      for (let pointIdx = 0; pointIdx < ring.length; pointIdx++) {
        try {
          const { lon, lat } = validatePointCoordinate(ring[pointIdx])
          allPoints.push({ lon, lat })
        } catch (e) {
          throw new Error(
            `Invalid coordinate at ring/line ${ringIdx}, point ${pointIdx}: ${e.message}`
          )
        }
      }
    }
    
    return allPoints
  } else if (expectedDepth === 3) {
    // MultiPolygon: [[[[lon, lat], ...], ...], ...]
    if (coordinates.length === 0) {
      throw new Error(`${geometryType} must have at least one polygon`)
    }
    
    const allPoints = []
    for (let polyIdx = 0; polyIdx < coordinates.length; polyIdx++) {
      const polygon = coordinates[polyIdx]
      if (!Array.isArray(polygon)) {
        throw new Error(`${geometryType} polygon at index ${polyIdx} must be an array`)
      }
      if (polygon.length === 0) {
        throw new Error(`${geometryType} polygon at index ${polyIdx} must have at least one ring`)
      }
      
      for (let ringIdx = 0; ringIdx < polygon.length; ringIdx++) {
        const ring = polygon[ringIdx]
        if (!Array.isArray(ring)) {
          throw new Error(
            `${geometryType} polygon ${polyIdx}, ring ${ringIdx} must be an array`
          )
        }
        if (ring.length === 0) {
          throw new Error(
            `${geometryType} polygon ${polyIdx}, ring ${ringIdx} must have at least one coordinate`
          )
        }
        
        for (let pointIdx = 0; pointIdx < ring.length; pointIdx++) {
          try {
            const { lon, lat } = validatePointCoordinate(ring[pointIdx])
            allPoints.push({ lon, lat })
          } catch (e) {
            throw new Error(
              `Invalid coordinate at polygon ${polyIdx}, ring ${ringIdx}, point ${pointIdx}: ${e.message}`
            )
          }
        }
      }
    }
    
    return allPoints
  } else {
    throw new Error(`Unsupported coordinate depth: ${expectedDepth}`)
  }
}

/**
 * Validate coordinates array matches the expected structure for the geometry type.
 * 
 * Validates:
 * 1. Structure depth matches geometry type
 * 2. All coordinates are within bounds (lon: -180 to 180, lat: -90 to 90)
 * 3. Coordinates are not swapped (lat/lon order)
 * 
 * @param {*} coordinates - Coordinate array to validate
 * @param {string} geometryType - Geometry type (Point, LineString, Polygon, MultiPoint, MultiLineString, MultiPolygon)
 * @returns {{valid: boolean, error: string|null}} - Validation result
 */
export function validateCoordinates(coordinates, geometryType) {
  try {
    // Reject null or undefined
    if (coordinates === null || coordinates === undefined) {
      return {
        valid: false,
        error: 'Coordinates cannot be null or empty'
      }
    }
    
    // Reject non-arrays
    if (!Array.isArray(coordinates)) {
      return {
        valid: false,
        error: `Coordinates must be an array, got ${typeof coordinates}`
      }
    }
    
    const geomType = (geometryType || '').toLowerCase()
    
    // Determine expected depth
    const depthMap = {
      'point': 0,
      'linestring': 1,
      'polygon': 2,
      'multipoint': 1,
      'multilinestring': 2,
      'multipolygon': 3,
    }
    
    if (!(geomType in depthMap)) {
      return {
        valid: false,
        error: `Unsupported geometry type: ${geometryType}`
      }
    }
    
    const expectedDepth = depthMap[geomType]
    
    // Reject empty arrays
    if (expectedDepth === 0 && coordinates.length === 0) {
      return {
        valid: false,
        error: 'Point coordinates cannot be empty. Must be [longitude, latitude]'
      }
    }
    
    if (expectedDepth > 0 && coordinates.length === 0) {
      return {
        valid: false,
        error: `${geometryType} coordinates cannot be empty`
      }
    }
    
    // Validate structure and get all points
    const points = validateCoordinatesStructure(coordinates, expectedDepth, geometryType)
    
    // Additional swap detection for multi-point geometries
    if (points.length > 1) {
      const swapError = checkMultiplePointsForSwap(points)
      if (swapError) {
        return {
          valid: false,
          error: swapError
        }
      }
    }
    
    return {
      valid: true,
      error: null
    }
  } catch (e) {
    return {
      valid: false,
      error: e.message
    }
  }
}

